### PR TITLE
Data path continued

### PR DESCRIPTION
**Pull request by ashgillman**
Progress on #89:

  - [x] Fix remaining PET Notebooks
  - [x] `chdir` is absolute so can be run multiple times
  - [x] `from sirf_exercises import exercises_data_path`
  - [x] `examples_data_path` -> `examples_data_path()`
  - [x] exercise_data_path.py flagged by git - this was because .gitingore is only for new files, not modified - now we don't modify a file but create
  - [ ] Separate working directories from data_directories

### CODE REVIEW

**Comment by ashgillman** [comment]
Testing completed:

ran `./scripts/download_data.sh`

ran `./scripts/download_data.sh data`

ran `./scripts/download_data.sh data ~/tmp` after copying downloaded files to `~/tmp` - they are found correctly

ran all modified notebooks until the `cd` command and ensured it ran

**Comment by ashgillman** [comment]
I've also tested the PET notebooks. MAPEM runs, but gives some funky results - I presume unrelated.

image

**Comment by ashgillman** [comment]
To explain how this works:

- each Notebook dir has a `notebook_setup.py` which need to be imported, and which will set the sys.path to point to SIRF-Exercises/lib
- Under lib are the library directories. If we wanted, these could include plotting and other repeated functions
- download_data.sh now creates lib/sirf_exercises/data_path.py with the location
- `exercises_data_path()` looks for that file, and also checks an ENV var which could be useful in the future but currently never really would be used. If it can't find the path it raises an exception.

File: scripts/download_data.sh [hunk 1]
```diff
@@ -35,10 +37,15 @@ while getopts 'pmoh' flag; do
...
+DOWNLOAD_DIR=${@:$OPTIND+1:2}              # parse optional DOWNLOAD_DIR
 SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
 REPO_DIR="$(dirname ${SCRIPT_DIR})"
-DATA_PATH=${DEST_DIR:-${REPO_DIR}/data}  # if no DEST_DIR, use REPO_DIR/data
+DATA_PATH=${DEST_DIR:-${REPO_DIR}/data}        # if no DEST_DIR, use REPO_DIR/data
+DOWNLOAD_DIR=${DOWNLOAD_DIR:-${DATA_PATH}}     # if no DOWNLOAD_DIR, use DEST_DIR
+DATA_PATH="$(readlink -f ${DATA_PATH})"        # canonicalise
+DOWNLOAD_DIR="$(readlink -f ${DOWNLOAD_DIR})"  # canonicalise
 echo Destination is ${DATA_PATH}
+echo $DOWNLOAD_DIR
```
Comment (KrisThielemans) [inline scripts/download_data.sh:48]:
```suggestion
echo Download location is \"${DOWNLOAD_DIR}\"
```

File: scripts/download_data.sh [hunk 2]
```diff
@@ -35,10 +37,15 @@ while getopts 'pmoh' flag; do
...
 
 DEST_DIR=${@:$OPTIND:1}                  # parse optional DEST_DIR
+DOWNLOAD_DIR=${@:$OPTIND+1:2}              # parse optional DOWNLOAD_DIR
 SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
 REPO_DIR="$(dirname ${SCRIPT_DIR})"
-DATA_PATH=${DEST_DIR:-${REPO_DIR}/data}  # if no DEST_DIR, use REPO_DIR/data
+DATA_PATH=${DEST_DIR:-${REPO_DIR}/data}        # if no DEST_DIR, use REPO_DIR/data
+DOWNLOAD_DIR=${DOWNLOAD_DIR:-${DATA_PATH}}     # if no DOWNLOAD_DIR, use DEST_DIR
+DATA_PATH="$(readlink -f ${DATA_PATH})"        # canonicalise
+DOWNLOAD_DIR="$(readlink -f ${DOWNLOAD_DIR})"  # canonicalise
```
Comment (KrisThielemans) [inline scripts/download_data.sh:46]: I don't think MacOS has `readlink`, https://stackoverflow.com/questions/1055671/how-can-i-get-the-behavior-of-gnus-readlink-f-on-a-mac
I'd cut it (unless we really need this?)

File: scripts/download_data.sh [hunk 3]
```diff
@@ -90,8 +97,8 @@ if [[ -n ${DO_PET} ]]
 then
     echo Downloading PET data
 
-    mkdir -p ${DATA_PATH}/PET/mMR
-    pushd ${DATA_PATH}/PET/mMR
+    mkdir -p ${DOWNLOAD_DIR}
```
Comment (KrisThielemans) [inline scripts/download_data.sh:100]: you could replace this with 
```suggestion
    mkdir -p "${DOWNLOAD_DIR}"
```
to be safe for spaces. or with a few less characters `"$DOWNLOAD_DIR"`.

or leave it for another day

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
looks good. I didn't try it but you did!

some interesting indentation choices going on? I guess we need a clangformat here as well :-) Just choose whatever was used before for now

**Comment by KrisThielemans** [comment]
also a codacy suggestion for quoting and spaces

File: scripts/download_data.sh (hunk 1 above)
Comment (ashgillman) [reply scripts/download_data.sh:48]: Whoops, that slipped in

File: scripts/download_data.sh [hunk 4]
```diff
@@ -23,63 +23,75 @@ print_usage() {
...
-  DO_OLD='true'
+REPO_DIR="$(dirname "${SCRIPT_DIR}")"
+DATA_PATH="${DEST_DIR:-"${REPO_DIR}"/data}"   # if no DEST_DIR, use REPO_DIR/data
+DOWNLOAD_DIR="${DOWNLOAD_DIR:-"$DATA_PATH"}"  # if no DOWNLOAD_DIR, use DEST_DIR
+DATA_PATH="$(canonicalise "$DATA_PATH")"        # canonicalise
+DOWNLOAD_DIR="$(canonicalise "$DOWNLOAD_DIR")"  # canonicalise
+echo Destination is \""$DATA_PATH"\"
+echo Download location is \""$DOWNLOAD_DIR"\"
+
+if ! [[ "$DO_PET" || "$DO_MR" || "$"DO_OLD ]]; then
```
Comment (KrisThielemans) [inline scripts/download_data.sh:62]:
```suggestion
if ! [[ $DO_PET || $DO_MR || $DO_OLD ]]; then
```
no quotes necessary as never any spaces? In any case, one quote pair was wrong

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
quotes everywhere 😄 !

I think now in too many places, but fine. I spotted only one mistake but am hoping you ran it again to see if it works.

**Comment by ashgillman** [comment]
Yes, I know I went overboard there... It was easier than checking line-by-line with Codeacy
I think it should still be valid

Huh, I did test.. I think that slipped through based on short-circuiting. I've tested with different flags a little more rigorously.
Although I also removed those particular line as it is now desirable to be able to run the script without downloading any data, if no flags are specified, if you wanted to just configure the path file.

I have no idea how to please Codeacy! If I put the function docstring to start after \``` it says Multi-line docstring summary should start at the second line (D212)" and if I put a newline it says "Multi-line docstring summary should start at the first line (D212)"!

**Comment by ashgillman** [comment]
Hmm, it has conflicting rules!
https://stackoverflow.com/questions/45990301/pep257-d212-and-d213-conflicts

**Comment by ashgillman** [comment]
I personally prefer

```
def some_func():
   """A docstring.

    more detail.
    """
    return
```

which would mean disabling D213

**Review by KrisThielemans — APPROVED** [review]

**Comment by KrisThielemans** [comment]
> Hmm, it has conflicting rules!
> https://stackoverflow.com/questions/45990301/pep257-d212-and-d213-conflicts

yes. I think the idea is that we choose one to enforce style. anyway

**Pull request merged**